### PR TITLE
refactor: Merge `status` and `consensus_status` endpoints

### PR DESCRIPTION
**Pull request by douglaz**
https://github.com/fedimint/fedimint/pull/2375 original goal was to add all the information to a single `status` endpoint, but at the time no good solution was found on how to merge both data. While creating the CLI command this simple and obvious solution appeared:
```rust
pub struct StatusResponse {
    pub server: ServerStatus,
    pub consensus: Option<ConsensusStatus>,
}
```

For instance in json:
```json
{
    "consensus": {
      "last_contribution": 14,
      "peers_flagged": 0,
      "peers_offline": 0,
      "peers_online": 3,
      "status_by_peer": {
        "1": {
          "connection_status": "Connected",
          "flagged": false,
          "last_contribution": 14,
          "last_contribution_timestamp_seconds": 1683215023
        },
        "2": {
          "connection_status": "Connected",
          "flagged": false,
          "last_contribution": 14,
          "last_contribution_timestamp_seconds": 1683215023
        },
        "3": {
          "connection_status": "Connected",
          "flagged": false,
          "last_contribution": 14,
          "last_contribution_timestamp_seconds": 1683215023
        }
      }
    },
    "server": "ConsensusRunning"
  }
```

### CODE REVIEW

File: fedimint-server/src/config/api.rs [hunk 1]
```diff
@@ -573,12 +574,13 @@ pub fn server_endpoints() -> Vec<ApiEndpoint<ConfigGenApi>> {
...
-                if context.has_auth() {
-                    Ok(config.status().await)
-                } else {
-                    Err(ApiError::unauthorized())
-                }
+            async |config: &ConfigGenApi, _context, _v: ()| -> StatusResponse {
+                let status = config.status().await;
+                assert_ne!(status, ServerStatus::ConsensusRunning);
+                Ok(StatusResponse {
+                    status,
```
Comment (justinmoon) [inline fedimint-server/src/config/api.rs:581]: Maybe this key should be `config_status` instead of just `status`?

Comment (douglaz) [reply fedimint-server/src/config/api.rs:581]: I think a more long name would something like `{server,general,overall}_status`. It's not specific to config. For instance in future it could say that we are in a `shutdown`/`terminating` state.

But I would like to hear @jkitman  thoughts on that.

Comment (jkitman) [reply fedimint-server/src/config/api.rs:581]: I would say that status isn't necessarily specific to config gen, just most of the current states are.

For instance we have a `ServerStatus::Upgrade` and `ServerStatus::ConsensusRunning` status.

The only issue with sticking the `Option<ConsensusStatus>` into this struct, is we might eventually need a `Option<UpgradeStatus>` or many other options.  But since this is mostly consumed by the UI, I'll leave it up to @justinmoon to decide whether this seems alright.

Comment (douglaz) [reply fedimint-server/src/config/api.rs:581]: I think it makes perfect sense to add a new field with a new specific struct when there is more information about a particular status.

File: fedimint-server/src/config/api.rs [hunk 2]
```diff
@@ -573,12 +574,13 @@ pub fn server_endpoints() -> Vec<ApiEndpoint<ConfigGenApi>> {
...
             "status",
-            async |config: &ConfigGenApi, context, _v: ()| -> ServerStatus {
-                if context.has_auth() {
-                    Ok(config.status().await)
-                } else {
-                    Err(ApiError::unauthorized())
-                }
+            async |config: &ConfigGenApi, _context, _v: ()| -> StatusResponse {
+                let status = config.status().await;
+                assert_ne!(status, ServerStatus::ConsensusRunning);
```
Comment (jkitman) [inline fedimint-server/src/config/api.rs:579]: I would not put asserts into production code

File: fedimint-core/src/api.rs [hunk 3]
```diff
@@ -905,6 +905,32 @@ pub struct ConsensusContribution {
...
+    VerifyingConfigs,
+    /// Restarted from a planned upgrade (requires action to start)
+    Upgrading,
+    /// Consensus is running
+    ConsensusRunning,
+}
+
+#[derive(Debug, Serialize, PartialEq, Eq)]
+pub struct StatusResponse {
+    pub status: ServerStatus,
```
Comment (dpc) [inline fedimint-core/src/api.rs:930]: The whole struct/endpoint is called "status", so it doesn't seem necessary  to repeat that word?

`server: ServerStatus` and `consensus:Option<ConsensusStatus>`?

File: fedimint-server/src/config/api.rs (hunk 1 above)
Comment (justinmoon) [reply fedimint-server/src/config/api.rs:580]: I'm good with `status`

Comment (douglaz) [reply fedimint-server/src/config/api.rs]: I think @dpc's suggestion make sense, use `server` for `ServerStatus` and `consensus` for `ConsensusStatus`

File: fedimint-core/src/api.rs [hunk 4]
```diff
@@ -905,6 +905,32 @@ pub struct ConsensusContribution {
...
+    Upgrading,
+    /// Consensus is running
+    ConsensusRunning,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
+pub struct StatusResponse {
+    pub server: ServerStatus,
+    pub consensus: Option<ConsensusStatus>,
+}
```
Comment (elsirion) [inline fedimint-core/src/api.rs:932]: Nit: Would it make sense to eventually merge these two structs? E.g. `ConsensusStatus` could be a member of `ServerStatus::ConsensusRunning`. Other states could also carry additional info (e.g. error messages for config gen failed).

File: fedimint-server/src/config/api.rs [hunk 5]
```diff
@@ -573,12 +574,12 @@ pub fn server_endpoints() -> Vec<ApiEndpoint<ConfigGenApi>> {
...
-                    Ok(config.status().await)
-                } else {
-                    Err(ApiError::unauthorized())
-                }
+            async |config: &ConfigGenApi, _context, _v: ()| -> StatusResponse {
+                let server = config.server_status().await;
+                Ok(StatusResponse {
+                    server,
+                    consensus: None
+                })
```
Comment (elsirion) [inline fedimint-server/src/config/api.rs:582]: I think we'll want to reconsider making this a public endpoint eventually, but ok for now.

**Review by elsirion — APPROVED** [review]

File: fedimint-core/src/api.rs (hunk 4 above)
Comment (douglaz) [reply fedimint-core/src/api.rs:932]: This was tried and the resulting serialized json is very hard to parse. I works if you're only using rust, but if you are interoperating with other system you need to simplify the result.
